include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +139,976 @@ struct Basic
-  {
-    quotient_and_pdf<SpectralBins> retval;
-    retval.quotient = _quotient;
-    retval.pdf = _pdf;
-    return retval;
-  }
-
-  SpectralBins value()
-  {
-    return quotient*pdf;
-  }
-  
-  SpectralBins quotient;
-  float pdf;
+    using this_t = quotient_and_pdf<SpectralBins, Pdf>;
+    static this_t create(NBL_CONST_REF_ARG(SpectralBins) _quotient, NBL_CONST_REF_ARG(Pdf) _pdf)
+    {
+        this_t retval;
+        retval.quotient = _quotient;
+        retval.pdf = _pdf;
+        return retval;
+    }
+
+    SpectralBins value()
+    {
+        return quotient*pdf;
+    }
+
+    SpectralBins quotient;
+    Pdf pdf;
+};
+
+typedef quotient_and_pdf<float32_t, float32_t> quotient_and_pdf_scalar;
+typedef quotient_and_pdf<vector<float32_t, 3>, float32_t> quotient_and_pdf_rgb;
+
+
+#define NBL_CONCEPT_NAME BxDF
+#define NBL_CONCEPT_TPLT_PRM_KINDS (typename)
+#define NBL_CONCEPT_TPLT_PRM_NAMES (T)
+#define NBL_CONCEPT_PARAM_0 (bxdf, T)
+#define NBL_CONCEPT_PARAM_1 (spec, typename T::spectral_type)
+#define NBL_CONCEPT_PARAM_2 (pdf, typename T::scalar_type)
+#define NBL_CONCEPT_PARAM_3 (_sample, typename T::sample_type)
+#define NBL_CONCEPT_PARAM_4 (iso, typename T::isotropic_type)
+#define NBL_CONCEPT_PARAM_5 (aniso, typename T::anisotropic_type)
+#define NBL_CONCEPT_PARAM_6 (param, typename T::params_t)
+NBL_CONCEPT_BEGIN(7)
+#define bxdf NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_0
+#define spec NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_1
+#define pdf NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_2
+#define _sample NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_3
+#define iso NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_4
+#define aniso NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_5
+#define param NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_6
+NBL_CONCEPT_END(
+    ((NBL_CONCEPT_REQ_TYPE)(T::scalar_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::isotropic_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::anisotropic_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::sample_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::spectral_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::quotient_pdf_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::params_t))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((bxdf.eval(param)), ::nbl::hlsl::is_same_v, typename T::scalar_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((bxdf.generate(aniso,aniso.N)), ::nbl::hlsl::is_same_v, typename T::sample_type))
+    //((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((T::template pdf<LS,I>(_sample,iso)), ::nbl::hlsl::is_scalar_v))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((bxdf.quotient_and_pdf(param)), ::nbl::hlsl::is_same_v, typename T::quotient_pdf_type))
+    ((NBL_CONCEPT_REQ_TYPE_ALIAS_CONCEPT)(Sample, typename T::sample_type))
+    ((NBL_CONCEPT_REQ_TYPE_ALIAS_CONCEPT)(spectral_of, typename T::spectral_type, typename T::scalar_type))
+    ((NBL_CONCEPT_REQ_TYPE_ALIAS_CONCEPT)(surface_interactions::Isotropic, typename T::isotropic_type))
+    ((NBL_CONCEPT_REQ_TYPE_ALIAS_CONCEPT)(surface_interactions::Anisotropic, typename T::anisotropic_type))
```

Comment:
wondering if we want out BxDFs to support lots of polymoprhism (every method templated), or be polymorphic themselves

Nothing to do right now, just an open question

Reply:
lots of parameter types need to be kept in sync, definitely template the whole struct not individual methods, if too many template args we can make a config meta type with aliases like the workgroup scan config struct